src/dbnode/encoding/proto/encoder.go

```diff
@@ -123,7 +125,7 @@ func (enc *Encoder) Encode(dp ts.Datapoint, timeUnit xtime.Unit, protoBytes ts.A
 	}
 
 	if enc.schema == nil {
-		return errEncoderSchemaIsRequired
+		panic(errEncoderSchemaIsRequired.Error())
```

Comment:
In this case, we really need it. I will add a comment.

It is a program error instead of a recoverable request error if schema is not supplied/set before encode is called. It means we missed some code path (to set schema).

Reply:
As long as the error is propagated properly it should come up fairly quickly if we missed something, but I guess its fair for now while this is all experimental, can always remove it later when we have more confidence

Reply:
agree, I wanted to flush all the bugs regarding schema not set asap, hence panic. We will remove it (add an error) later after we gain confidence.

Reply:
Hm I feel pretty strongly this should be a `return err` not a panic.

Reply:
Can you use `instrument.InvariantErrorf(...)` here?  And then you can use an environment variable to make it panic.

See this code:
https://github.com/m3db/m3/blob/master/src/x/instrument/invariant.go#L83-L91

Which will panic if run with:
```
PANIC_ON_INVARIANT_VIOLATED=true ./m3dbnode -f ....
```